parsec/core/fs/workspacefs/workspacefile.py

```diff
@@ -0,0 +1,234 @@
+
+
+class FileState(Enum):
+    INIT = 0
+    OPEN = 1
+    CLOSED = 2
+
+
+def check_state(func):
+    @functools.wraps(func)
+    def wrapper(self, *args, **kwargs):
+        if self.closed or self._state == FileState.INIT:
+            raise ValueError("I/O operation on closed file.")
+
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+class WorkspaceFile:
+    """Async file Object.
+
+    Methods of this object should act like trio async file :
+    "https://github.com/python-trio/trio/blob/master/trio/_file_io.py" using transactions.
+
+    If you are using this object outside a context manager you should call ainit() method first.
+
+    Keyword arguments:
+    fd -- the file descriptor.
+    transactions -- object to use for file transactions.
+    path -- relative path of the file.
+    mode -- the mode where the file have been opened.
+    mode("a") -> append file.
+    mode("b") -> bytes, have to be used with another mode, ignored, class is already working with
+    bytes.
+    mode("w") -> write file.
+    mode("r") -> read file.
+    """
+
+    def __init__(self, fd: int, transactions, path: FsPath, mode: str = "r"):
+        self._fd = fd
+        self._offset = 0
+        self._state = FileState.INIT
+        self._path = path
+        self._transactions = transactions
+        mode = mode.lower()
+        # Preventing to open in write and read in same time or write and append or open with no mode
+        if sum(c in mode for c in "rwa") != 1:
+            raise ValueError("must have exactly one of create/read/write/append mode")
+        # Preventing to open with non-existant mode
+        elif re.search("[^arwb]", mode) is not None:
+            raise ValueError("invalid mode: ", mode)
+        self._mode = mode
+
+    async def __aenter__(self):
+        await self.ainit()
+        return self
+
+    async def __aexit__(self, *args):
+        self._state = FileState.CLOSED
+        await self.close()
+
+    async def ainit(self):
+        """Initializing the File Object.
+
+            Check the FileState and truncate the file if needed.
+        """
+        if self._state == FileState.INIT:
+            self._state = FileState.OPEN
+            # Checking if the 'w' open mode have been selected, if yes, truncate the file
+            if "w" in self._mode:
+                await self.truncate(0)
+
+    async def __anext__(self):
+        return await self.readline()
+
+    async def close(self):
+        """Close the file"""
+        if self._state != FileState.CLOSED:
+            self._state = FileState.CLOSED
+            await self._transactions.fd_close(self._fd)
+
+    @property
+    def closed(self):
+        return self._state == FileState.CLOSED
+
+    @check_state
+    async def file_stat(self):
+        """Getting file stat"""
+        stats = await self._transactions.fd_info(self._fd, self._path)
+        return stats
+
+    @check_state
+    async def get_filelength(self):
```

Comment:
I think `size` is a better name than `length`, maybe `get_size()` or `file_size()`?